File: src/main/java/de/blazemcworld/fireflow/command/VariablesCommand.java
package de.blazemcworld.fireflow.command;

import com.mojang.brigadier.Command;
import com.mojang.brigadier.arguments.StringArgumentType;
import com.mojang.brigadier.builder.LiteralArgumentBuilder;
import de.blazemcworld.fireflow.code.VariableStore;
import de.blazemcworld.fireflow.space.Space;
import net.minecraft.commands.CommandSourceStack;
import net.minecraft.commands.Commands;
import net.minecraft.network.chat.Component;
import net.minecraft.network.chat.Style;
import net.minecraft.network.chat.TextColor;
import net.minecraft.server.level.ServerPlayer;

import java.util.Set;
import java.util.function.Predicate;

public class VariablesCommand {

    public static void attach(LiteralArgumentBuilder<CommandSourceStack> node) {
        attach(node, "variables");
        attach(node, "vars");
    }

    private static void attach(LiteralArgumentBuilder<CommandSourceStack> node, String alias) {
        node.then(Commands.literal(alias)
                .executes(ctx -> {
                    listVariables(ctx.getSource(), null);
                    return Command.SINGLE_SUCCESS;
                })
                .then(Commands.argument("filter", StringArgumentType.greedyString())
                        .executes(ctx -> {
                            listVariables(ctx.getSource(), StringArgumentType.getString(ctx, "filter"));
                            return Command.SINGLE_SUCCESS;
                        })
                )
        );
    }

    private static void listVariables(CommandSourceStack source, String query) {
        ServerPlayer player = CommandHelper.getPlayer(source);
        Space space = CommandHelper.getSpace(player);
        if (!CommandHelper.isDeveloperOrOwner(player, space)) return;

        String lowerQuery = query == null ? null : query.toLowerCase();
        Predicate<String> filter = query == null ? s -> true : s -> s.toLowerCase().contains(lowerQuery);

        Set<VariableStore.VarEntry> vars = space.savedVariables.iterator(filter, 50);
        for (VariableStore.VarEntry v : vars) {
            player.sendSystemMessage(Component.literal(v.name()).setStyle(Style.EMPTY.withColor(v.type().color))
                    .append(Component.literal(": ").withColor(TextColor.GRAY))
                    .append(Component.literal(v.type().stringify(v.value(), "display")).withColor(TextColor.WHITE)));
        }

        if (vars.size() >= 50) return;

        for (VariableStore.VarEntry v : space.evaluator.sessionVariables.iterator(filter, 50 - vars.size())) {
            player.sendSystemMessage(Component.literal(v.name()).setStyle(Style.EMPTY.withColor(v.type().color))
                    .append(Component.literal(": ").withColor(TextColor.GRAY))
                    .append(Component.literal(v.type().stringify(v.value(), "display")).withColor(TextColor.WHITE)));
        }
    }

}

File: src/main/kotlin/de/blazemcworld/fireflow/command/VariablesCommand.kt
package de.blazemcworld.fireflow.command

import de.blazemcworld.fireflow.database.table.PlayersTable
import de.blazemcworld.fireflow.database.table.SpaceRolesTable
import de.blazemcworld.fireflow.node.ValueType
import de.blazemcworld.fireflow.space.SpaceManager
import de.blazemcworld.fireflow.util.sendError
import net.kyori.adventure.text.Component
import net.minestom.server.command.builder.Command
import net.minestom.server.command.builder.CommandExecutor
import net.minestom.server.command.builder.arguments.ArgumentString
import net.minestom.server.entity.Player
import org.jetbrains.exposed.sql.JoinType
import org.jetbrains.exposed.sql.SqlExpressionBuilder.eq
import org.jetbrains.exposed.sql.and
import org.jetbrains.exposed.sql.selectAll
import org.jetbrains.exposed.sql.transactions.transaction

object VariablesCommand : Command("variables", "vars") {
    init {
        defaultExecutor = CommandExecutor exec@{ sender, ctx ->
            if (sender !is Player) return@exec

            val space = SpaceManager.currentSpace(sender)

            if (space == null) {
                sender.sendError("You need to be in a space to do this!")
                return@exec
            }

            val role = transaction {
                val result = SpaceRolesTable.join(PlayersTable, JoinType.INNER, SpaceRolesTable.player, PlayersTable.id)
                    .selectAll().where((SpaceRolesTable.space eq space.id) and (PlayersTable.uuid eq sender.uuid))
                    .adjustSelect { select(SpaceRolesTable.role) }
                if (result.empty()) return@transaction null
                result.single()[SpaceRolesTable.role]
            }

            if (role != SpaceRolesTable.Role.OWNER && role != SpaceRolesTable.Role.CONTRIBUTOR) {
                sender.sendError("You are not allowed to do that!")
                return@exec
            }

            var msgs = 0
            for ((name, value) in space.varStore) {
                if (!name.contains(ctx.get("filter") ?: "")) continue
                sender.sendMessage(Component.text(name + " -> " + (value.first as ValueType<Any>).stringify(value.second)).color(value.first.color))
                if (msgs++ > 100) break
            }
        }

        addSyntax(defaultExecutor!!, ArgumentString("filter"))
    }
}